### PR TITLE
Cache results of `forward` and introduce a super class for ML models.

### LINKED ISSUE #58: Compute forward pass only once. Re-use the results of forward in backward
### What is the expected enhancement?
Now in `NeuralNetworkClassifier/Regressor` forward pass is computed at least two times, in `objective` and `objective_grad`. This should be changed in a way forward pass is computed only once and the results are stored for re-use in the backward pass.

### PR DESCRIPTION
**Pull request by adekusar-drl**
### Summary
This PR closes #58 by caching `forward` results and introduces a base class `TrainableModel` for ML models and `ObjectiveFunction` as a base class for objective functions optimized in classifier/regressor.

### Details and comments
Since classes `NeuralNetworkClassifier` and `NeuralNetworkRegressor` optimize similar objective functions and the goal of this PR is to compute `forward` pass only once, it is reasonable to pull out these objective functions of the classifier/regressor and add an ability to cache last results of `forward` in the base class `ObjectiveFunction` of the objective functions. In the same way we introduce a base class `TrainableModel` for ML models (classifier and regressor) to reduce the amount of common code and place this common code in the base model class. Names of the base classes are open for discussion.

New classes in the PR:

- `TrainableModel` a base/super class for the NN classifier and regressor. Both of extend this class. This class defines Scikit-Learn interface.
- `ObjectiveFunction` for optimization in the ML models. It has `objective` and `gradient` methods. There are also `BinaryObjectiveFunction`, `MultiClassObjectiveFunction`, and `OneHotObjectiveFunction`.

### CODE REVIEW

File: qiskit_machine_learning/algorithms/trainable_model.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        else:
+            loss = loss.lower()
+            if loss == "l1":
+                self._loss = L1Loss()
+            elif loss == "l2":
+                self._loss = L2Loss()
+            elif loss == "cross_entropy":
+                self._loss = CrossEntropyLoss()
+            elif loss == "cross_entropy_sigmoid":
+                self._loss = CrossEntropySigmoidLoss()
```
Comment (dlasecki) [inline qiskit_machine_learning/algorithms/trainable_model.py:76]: Shall we extract it to a separate class that would handle providing a correct Loss object based on the string? I can imagine that this string-based convention might reappear in many classes and then each of them needs this long block of if statements which leads to code duplication.

File: qiskit_machine_learning/algorithms/objective_functions.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+class OneHotObjectiveFunction(ObjectiveFunction):
+    """
+    An objective function for one hot encoding representation of the output,
+    e.g. classes like ``[1, 0, 0]``, ``[0, 1, 0]``, ``[0, 0, 1]``.
+    """
+
+    def objective(self, weights: np.ndarray) -> float:
+        val = 0.0
+        probs = self._neural_network_forward(weights)
+        # TODO: do batch eval
```
Comment (dlasecki) [inline qiskit_machine_learning/algorithms/objective_functions.py:146]: Was it intended to address the TODO in the current PR?

File: qiskit_machine_learning/algorithms/objective_functions.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        for i in range(len(self._X)):
+            val += self._loss(probs[i], self._y[i])
+        return val
+
+    def gradient(self, weights: np.ndarray) -> np.ndarray:
+        grad = np.zeros(self._neural_network.num_weights)
+        y_predict = self._neural_network_forward(weights)
+        # TODO: do batch eval for backward instead of zip()
```
Comment (dlasecki) [inline qiskit_machine_learning/algorithms/objective_functions.py:154]: Was it intended to address the TODO in the current PR?

File: qiskit_machine_learning/algorithms/objective_functions.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        probs = self._neural_network_forward(weights)
+        for i in range(len(self._X)):
+            for y_predict, prob in enumerate(probs[i]):
+                val += prob * self._loss(y_predict, self._y[i])
+        return val
+
+    def gradient(self, weights: np.ndarray) -> np.ndarray:
+        num_classes = self._neural_network.output_shape[0]
+        grad = np.zeros((1, self._neural_network.num_weights))
+        # TODO: do batch eval instead of zip()
```
Comment (dlasecki) [inline qiskit_machine_learning/algorithms/objective_functions.py:129]: Was it intended to address the TODO in the current PR?

File: qiskit_machine_learning/algorithms/trainable_model.py (hunk 1 above)
Comment (adekusar-drl) [reply qiskit_machine_learning/algorithms/trainable_model.py:76]: In general, I agree. But I'd rather wait till the time this situation occurs or address it when #82 will be in progress.

File: qiskit_machine_learning/algorithms/objective_functions.py (hunk 2 above)
Comment (adekusar-drl) [reply qiskit_machine_learning/algorithms/objective_functions.py:146]: No, there's issue #63 to review all of these TODOs.

File: qiskit_machine_learning/algorithms/objective_functions.py (hunk 4 above)
Comment (adekusar-drl) [reply qiskit_machine_learning/algorithms/objective_functions.py:129]: As above, there's #63.

**Comment by adekusar-drl** [comment]
> There are a number of new classes added here and the documentation, nor init file, have been updated to add them as docs or **all**. Shouldn't they be more exposed this way - although others to re-use them as needed etc. Any maybe a reno would be good as I imagine this is intended as a performance improvement

Agree, these are the things I usually forget to do.

File: qiskit_machine_learning/algorithms/objective_functions.py [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+        Computes and caches the results of the forward pass. Cached values may be re-used in
+        gradient computation.
+
+        Args:
+            weights: an array of weights to be used in the forward pass.
+
+        Returns:
+            The result of the neural network.
+        """
+        data_id = str(id(weights)) + str(id(self._X))
```
[the remark this answers is not in the document]
Comment (adekusar-drl) [reply qiskit_machine_learning/algorithms/objective_functions.py:84]: I'd expect a comparison by id to be faster than a direct comparison of arrays. Well, it is not the case now, we have just a few qubits, but in case of classic ML, arrays maybe huge.
It is also a good point whether we need to add `_X` to the comparison. Honestly, I added it only for safety reasons. In general, a comparison of weights is good enough, but just in case. So, if you don't mind I'd keep it.

[the remark this answers is not in the document]
Comment (woodsp-ibm) [reply qiskit_machine_learning/algorithms/objective_functions.py:84]: Sure, I was just trying to understand things from whats there. Whether its worth a comment in the code to that effect, i.e. safety just in case, to help any other new readers of the code I'll leave to you.

**Review by woodsp-ibm — APPROVED** [review]
